File: scripts/capsule_runtime_health.py
#!/usr/bin/env python3
"""Read-only capsule install/discovery/runtime compatibility health probe."""

from __future__ import annotations

import argparse
import hashlib
import json
import os
import sys
import tempfile
import textwrap
import tomllib
import unittest
from pathlib import Path
from typing import Any


ASTRID_REPO = Path(__file__).resolve().parents[1]
DEFAULT_BASELINE = ASTRID_REPO / "scripts/baselines/capsule_runtime_health.json"

CORE_WASM_VERSION = b"\x01\x00\x00\x00"
WASM_MAGIC = b"\x00asm"


def astrid_home() -> Path:
    return Path(os.environ.get("ASTRID_HOME", str(Path.home() / ".astrid")))


def manifest_paths(capsules_dir: Path) -> list[Path]:
    paths: list[Path] = []
    direct = capsules_dir / "Capsule.toml"
    if direct.is_file():
        paths.append(direct)
    if capsules_dir.is_dir():
        for child in sorted(capsules_dir.iterdir()):
            manifest = child / "Capsule.toml"
            if child.is_dir() and manifest.is_file():
                paths.append(manifest)
    return paths


def discovery_dirs(repo: Path, home: Path) -> list[Path]:
    return [
        home / "home/default/.local/capsules",
        repo / ".astrid/capsules",
    ]


def load_json(path: Path) -> dict[str, Any]:
    try:
        data = json.loads(path.read_text())
    except Exception:
        return {}
    return data if isinstance(data, dict) else {}


def load_baseline(path: Path | None, disabled: bool) -> dict[str, Any]:
    if disabled or path is None or not path.is_file():
        return {
            "enabled": False,
            "path": str(path) if path else None,
            "accepted_legacy_extism_mvp": [],
        }
    data = load_json(path)
    accepted = data.get("accepted_legacy_extism_mvp", [])
    return {
        "enabled": True,
        "path": str(path),
        "accepted_legacy_extism_mvp": [
            item for item in accepted if isinstance(item, dict)
        ],
    }


def accepted_legacy(baseline: dict[str, Any], name: str, wasm_hash: str | None) -> bool:
    for item in baseline.get("accepted_legacy_extism_mvp", []):
        if item.get("name") != name:
            continue
        expected_hash = item.get("wasm_hash")
        if expected_hash is None or expected_hash == wasm_hash:
            return True
    return False


def parse_manifest(path: Path) -> dict[str, Any] | None:
    try:
        data = tomllib.loads(path.read_text())
    except Exception:
        return None
    return data if isinstance(data, dict) else None


def meta_wasm_hash(capsule_dir: Path) -> str | None:
    value = load_json(capsule_dir / "meta.json").get("wasm_hash")
    return value if isinstance(value, str) and value else None


def resolve_payload(
    capsule_dir: Path,
    component_path: str,
    home: Path,
    wasm_hash: str | None,
) -> Path | None:
    raw = Path(component_path)
    local = raw if raw.is_absolute() else capsule_dir / raw
    if local.is_file():
        return local
    if wasm_hash:
        hashed = home / "bin" / f"{wasm_hash}.wasm"
        if hashed.is_file():
            return hashed
    return None


def classify_wasm(path: Path) -> str:
    try:
        data = path.read_bytes()
    except Exception:
        return "missing_payload"
    if len(data) < 8 or data[:4] != WASM_MAGIC:
        return "invalid_wasm"
    if data[4:8] == CORE_WASM_VERSION:
        if b"extism" in data.lower():
            return "legacy_extism_mvp"
        return "core_module_mvp"
    return "component_model"


def short_hash(path: Path | None) -> str | None:
    if path is None or not path.is_file():
        return None
    return hashlib.sha256(path.read_bytes()).hexdigest()[:12]


def build_report(
    repo: Path = ASTRID_REPO,
    home: Path | None = None,
    baseline_path: Path | None = DEFAULT_BASELINE,
    no_baseline: bool = False,
) -> dict[str, Any]:
    home = home or astrid_home()
    baseline = load_baseline(baseline_path, no_baseline)
    dirs = discovery_dirs(repo, home)
    raw_paths = [path for directory in dirs for path in manifest_paths(directory)]

    capsules: list[dict[str, Any]] = []
    seen: set[str] = set()
    for manifest_path in raw_paths:
        manifest = parse_manifest(manifest_path)
        if not manifest:
            continue
        package = manifest.get("package", {})
        name = package.get("name")
        if not isinstance(name, str) or not name or name in seen:
            continue
        seen.add(name)

        components = manifest.get("component") or []
        if isinstance(components, dict):
            components = [components]
        component = components[0] if components and isinstance(components[0], dict) else None
        capsule_dir = manifest_path.parent
        wasm_hash = meta_wasm_hash(capsule_dir)
        payload = None
        component_path = None
        if component:
            raw_component_path = component.get("file") or component.get("entrypoint")
            if isinstance(raw_component_path, str):
                component_path = raw_component_path
                payload = resolve_payload(capsule_dir, component_path, home, wasm_hash)

        runtime_status = "no_component"
        baseline_status = "not_applicable"
        actionable = False
        if component:
            if payload is None:
                runtime_status = "missing_payload"
                baseline_status = "new"
                actionable = True
            else:
                runtime_status = classify_wasm(payload)
                if runtime_status == "component_model":
                    baseline_status = "current"
                elif runtime_status == "legacy_extism_mvp" and accepted_legacy(
                    baseline, name, wasm_hash
                ):
                    baseline_status = "accepted"
                else:
                    baseline_status = "new"
                    actionable = True

        capsules.append(
            {
                "name": name,
                "version": package.get("version"),
                "manifest_path": str(manifest_path),
                "component_path": component_path,
                "resolved_payload": str(payload) if payload else None,
                "wasm_hash": wasm_hash,
                "payload_sha256_12": short_hash(payload),
                "runtime_status": runtime_status,
                "baseline_status": baseline_status,
                "actionable": actionable,
            }
        )

    component_capsules = [c for c in capsules if c["runtime_status"] != "no_component"]
    legacy = [c for c in capsules if c["runtime_status"] == "legacy_extism_mvp"]
    accepted_legacy_items = [c for c in legacy if c["baseline_status"] == "accepted"]
    missing = [c for c in capsules if c["runtime_status"] == "missing_payload"]
    actionable_items = [c for c in capsules if c["actionable"]]
    summary = {
        "status": "ok" if not actionable_items else "warning",
        "installed_manifests": len(raw_paths),
        "discovered_manifests": len(capsules),
        "component_payloads_found": len(
            [c for c in component_capsules if c["resolved_payload"]]
        ),
        "loadable_component_model": len(
            [c for c in capsules if c["runtime_status"] == "component_model"]
        ),
        "legacy_extism_mvp": len(legacy),
        "accepted_legacy_extism_mvp": len(accepted_legacy_items),
        "missing_payloads": len(missing),
        "actionable_incompatible": len(
            [
                c
                for c in actionable_items
                if c["runtime_status"] != "missing_payload"
            ]
        ),
        "actionable_missing_payloads": len(
            [c for c in actionable_items if c["runtime_status"] == "missing_payload"]
        ),
    }
    return {
        "summary": summary,
        "capsules": capsules,
        "actionable_capsules": actionable_items,
        "accepted_legacy_extism_mvp": accepted_legacy_items,
        "baseline": {
            **baseline,
            "accepted_legacy_extism_mvp_count": len(
                baseline.get("accepted_legacy_extism_mvp", [])
            ),
        },
    }


def render_markdown(report: dict[str, Any], show_accepted: bool) -> str:
    summary = report["summary"]
    lines = [
        "# Capsule Runtime Health",
        "",
        f"- Status: `{summary['status']}`",
        f"- Installed/discovered manifests: `{summary['installed_manifests']}` / `{summary['discovered_manifests']}`",
        f"- Payloads found: `{summary['component_payloads_found']}`",
        f"- Component Model payloads: `{summary['loadable_component_model']}`",
        f"- Accepted legacy Extism/MVP backlog: `{summary['accepted_legacy_extism_mvp']}` / `{summary['legacy_extism_mvp']}`",
        f"- Actionable incompatible/missing: `{summary['actionable_incompatible']}` / `{summary['actionable_missing_payloads']}`",
        "",
    ]
    actionable = report.get("actionable_capsules", [])
    if actionable:
        lines.append("## Actionable")
        for item in actionable:
            lines.append(
                f"- `{item['name']}`: {item['runtime_status']} ({item.get('manifest_path')})"
            )
        lines.append("")
    if show_accepted:
        accepted = report.get("accepted_legacy_extism_mvp", [])
        lines.append("## Accepted Legacy Backlog")
        if accepted:
            for item in accepted:
                lines.append(f"- `{item['name']}`: {item['wasm_hash']}")
        else:
            lines.append("- None")
    return "\n".join(lines)


class CapsuleRuntimeHealthTests(unittest.TestCase):
    def write_capsule(
        self,
        root: Path,
        name: str,
        wasm_hash: str,
        wasm_bytes: bytes | None,
    ) -> None:
        capsule_dir = root / "home/default/.local/capsules" / name
        capsule_dir.mkdir(parents=True)
        (capsule_dir / "Capsule.toml").write_text(
            textwrap.dedent(
                f"""
                [package]
                name = "{name}"
                version = "0.1.0"

                [[component]]
                id = "main"
                file = "{name}.wasm"
                """
            ).strip()
        )
        (capsule_dir / "meta.json").write_text(json.dumps({"wasm_hash": wasm_hash}))
        if wasm_bytes is not None:
            bin_dir = root / "bin"
            bin_dir.mkdir(parents=True, exist_ok=True)
            (bin_dir / f"{wasm_hash}.wasm").write_bytes(wasm_bytes)

    def test_accepted_legacy_is_ok(self) -> None:
        with tempfile.TemporaryDirectory() as tmp:
            home = Path(tmp) / "home"
            repo = Path(tmp) / "repo"
            repo.mkdir()
            baseline = repo / "baseline.json"
            self.write_capsule(home, "legacy", "abc", WASM_MAGIC + CORE_WASM_VERSION + b"extism")
            baseline.write_text(
                json.dumps({"accepted_legacy_extism_mvp": [{"name": "legacy", "wasm_hash": "abc"}]})
            )
            report = build_report(repo=repo, home=home, baseline_path=baseline)
            self.assertEqual(report["summary"]["status"], "ok")
            self.assertEqual(report["summary"]["accepted_legacy_extism_mvp"], 1)

    def test_new_core_module_escalates(self) -> None:
        with tempfile.TemporaryDirectory() as tmp:
            home = Path(tmp) / "home"
            repo = Path(tmp) / "repo"
            repo.mkdir()
            self.write_capsule(home, "new-core", "def", WASM_MAGIC + CORE_WASM_VERSION)
            report = build_report(repo=repo, home=home, baseline_path=None)
            self.assertEqual(report["summary"]["status"], "warning")
            self.assertEqual(report["summary"]["actionable_incompatible"], 1)

    def test_missing_payload_escalates(self) -> None:
        with tempfile.TemporaryDirectory() as tmp:
            home = Path(tmp) / "home"
            repo = Path(tmp) / "repo"
            repo.mkdir()
            self.write_capsule(home, "missing", "ghi", None)
            report = build_report(repo=repo, home=home, baseline_path=None)
            self.assertEqual(report["summary"]["status"], "warning")
            self.assertEqual(report["summary"]["actionable_missing_payloads"], 1)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--json", action="store_true")
    parser.add_argument("--strict", action="store_true")
    parser.add_argument("--show-accepted", action="store_true")
    parser.add_argument("--no-baseline", action="store_true")
    parser.add_argument("--baseline", type=Path, default=DEFAULT_BASELINE)
    parser.add_argument("--repo", type=Path, default=ASTRID_REPO)
    parser.add_argument("--home", type=Path, default=None)
    parser.add_argument("--self-test", action="store_true")
    args = parser.parse_args()

    if args.self_test:
        suite = unittest.defaultTestLoader.loadTestsFromTestCase(CapsuleRuntimeHealthTests)
        result = unittest.TextTestRunner(verbosity=2).run(suite)
        return 0 if result.wasSuccessful() else 1

    report = build_report(
        repo=args.repo,
        home=args.home,
        baseline_path=args.baseline,
        no_baseline=args.no_baseline,
    )
    if args.json:
        print(json.dumps(report, indent=2, sort_keys=True))
    else:
        print(render_markdown(report, args.show_accepted))
    if args.strict and report["summary"]["status"] != "ok":
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: crates/astrid-kernel/src/capsule_runtime_health.rs
use std::collections::HashSet;
use std::path::{Path, PathBuf};

use astrid_events::kernel_api::CapsuleRuntimeHealth;

#[derive(Debug, Default, serde::Deserialize)]
struct Baseline {
    #[serde(default)]
    accepted_legacy_extism_mvp: Vec<BaselineEntry>,
}

#[derive(Debug, serde::Deserialize)]
struct BaselineEntry {
    name: String,
    #[serde(default)]
    wasm_hash: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum PayloadKind {
    ComponentModel,
    LegacyExtismMvp,
    CoreModuleMvp,
    Invalid,
}

pub(crate) fn summarize(workspace_root: &Path, loaded_capsules: &[String]) -> CapsuleRuntimeHealth {
    let dirs = discovery_dirs(workspace_root);
    let installed_manifests = count_manifest_paths(&dirs);
    let discovered = discover_manifest_entries(&dirs);
    let baseline = load_baseline(workspace_root).unwrap_or_default();
    let home_bin = astrid_core::dirs::AstridHome::resolve()
        .ok()
        .map(|home| home.bin_dir());

    let mut health = CapsuleRuntimeHealth {
        status: "ok".to_string(),
        installed_manifests: to_u32(installed_manifests),
        discovered_manifests: to_u32(discovered.len()),
        loaded_capsules: to_u32(loaded_capsules.len()),
        ..CapsuleRuntimeHealth::default()
    };

    for (manifest, dir) in discovered {
        let Some(component) = manifest.components.first() else {
            continue;
        };
        let wasm_hash = read_meta_wasm_hash(&dir);
        let Some(payload_path) =
            resolve_component_payload(&dir, &component.path, home_bin.as_ref())
        else {
            health.missing_payloads = health.missing_payloads.saturating_add(1);
            health.actionable_missing_payloads =
                health.actionable_missing_payloads.saturating_add(1);
            continue;
        };

        health.component_payloads_found = health.component_payloads_found.saturating_add(1);
        match classify_payload(&payload_path) {
            PayloadKind::ComponentModel => {
                health.loadable_component_model = health.loadable_component_model.saturating_add(1);
            },
            PayloadKind::LegacyExtismMvp => {
                health.legacy_extism_mvp = health.legacy_extism_mvp.saturating_add(1);
                if accepted_legacy(&baseline, &manifest.package.name, wasm_hash.as_deref()) {
                    health.accepted_legacy_extism_mvp =
                        health.accepted_legacy_extism_mvp.saturating_add(1);
                } else {
                    health.actionable_incompatible =
                        health.actionable_incompatible.saturating_add(1);
                }
            },
            PayloadKind::CoreModuleMvp | PayloadKind::Invalid => {
                health.actionable_incompatible = health.actionable_incompatible.saturating_add(1);
            },
        }
    }

    if health.actionable_incompatible > 0 || health.actionable_missing_payloads > 0 {
        health.status = "warning".to_string();
    }
    health
}

fn discovery_dirs(workspace_root: &Path) -> Vec<PathBuf> {
    let mut dirs = Vec::new();
    if let Ok(home) = astrid_core::dirs::AstridHome::resolve() {
        let principal = astrid_core::PrincipalId::default();
        dirs.push(home.principal_home(&principal).capsules_dir());
    }
    dirs.push(workspace_root.join(".astrid").join("capsules"));
    dirs
}

fn count_manifest_paths(dirs: &[PathBuf]) -> usize {
    dirs.iter().map(|dir| manifest_paths(dir).len()).sum()
}

fn discover_manifest_entries(
    dirs: &[PathBuf],
) -> Vec<(astrid_capsule::manifest::CapsuleManifest, PathBuf)> {
    let mut seen = HashSet::new();
    let mut entries = Vec::new();
    for manifest_path in dirs.iter().flat_map(|dir| manifest_paths(dir)) {
        let capsule_dir = manifest_path
            .parent()
            .map_or_else(|| PathBuf::from("."), Path::to_path_buf);
        if let Ok(manifest) = astrid_capsule::discovery::load_manifest(&manifest_path)
            && seen.insert(manifest.package.name.clone())
        {
            entries.push((manifest, capsule_dir));
        }
    }
    entries
}

fn manifest_paths(dir: &Path) -> Vec<PathBuf> {
    let mut paths = Vec::new();
    let direct = dir.join("Capsule.toml");
    if direct.is_file() {
        paths.push(direct);
    }
    let Ok(entries) = std::fs::read_dir(dir) else {
        return paths;
    };
    for entry in entries.flatten() {
        let path = entry.path();
        if path.is_dir() {
            let manifest = path.join("Capsule.toml");
            if manifest.is_file() {
                paths.push(manifest);
            }
        }
    }
    paths.sort();
    paths
}

fn read_meta_wasm_hash(capsule_dir: &Path) -> Option<String> {
    let meta_path = capsule_dir.join("meta.json");
    let meta: serde_json::Value = serde_json::from_slice(&std::fs::read(meta_path).ok()?).ok()?;
    meta.get("wasm_hash")?.as_str().map(String::from)
}

fn resolve_component_payload(
    capsule_dir: &Path,
    component_path: &Path,
    home_bin: Option<&PathBuf>,
) -> Option<PathBuf> {
    let local = if component_path.is_absolute() {
        component_path.to_path_buf()
    } else {
        capsule_dir.join(component_path)
    };
    if local.is_file() {
        return Some(local);
    }

    let hash = read_meta_wasm_hash(capsule_dir)?;
    let hashed = home_bin?.join(format!("{hash}.wasm"));
    hashed.is_file().then_some(hashed)
}

fn classify_payload(path: &Path) -> PayloadKind {
    let Ok(bytes) = std::fs::read(path) else {
        return PayloadKind::Invalid;
    };

    match wasmparser::Parser::new(0).parse_all(&bytes).next() {
        Some(Ok(wasmparser::Payload::Version {
            encoding: wasmparser::Encoding::Component,
            ..
        })) => PayloadKind::ComponentModel,
        Some(Ok(wasmparser::Payload::Version {
            encoding: wasmparser::Encoding::Module,
            ..
        })) => {
            if contains_extism_marker(&bytes) {
                PayloadKind::LegacyExtismMvp
            } else {
                PayloadKind::CoreModuleMvp
            }
        },
        _ => PayloadKind::Invalid,
    }
}

fn contains_extism_marker(bytes: &[u8]) -> bool {
    bytes
        .windows("extism".len())
        .any(|window| window.eq_ignore_ascii_case(b"extism"))
}

fn load_baseline(workspace_root: &Path) -> Option<Baseline> {
    let path = workspace_root
        .join("scripts")
        .join("baselines")
        .join("capsule_runtime_health.json");
    serde_json::from_slice(&std::fs::read(path).ok()?).ok()
}

fn accepted_legacy(baseline: &Baseline, name: &str, wasm_hash: Option<&str>) -> bool {
    baseline.accepted_legacy_extism_mvp.iter().any(|entry| {
        entry.name == name
            && match entry.wasm_hash.as_deref() {
                Some(expected) => wasm_hash == Some(expected),
                None => true,
            }
    })
}

fn to_u32(value: usize) -> u32 {
    u32::try_from(value).unwrap_or(u32::MAX)
}
